### PR TITLE
Geometry snapping on adaptivity

### LINKED ISSUE #9587: Snap to Geometry
### Description of the enhancement or error report
For curved domains it would be nice to have new nodes generated through adaptivity to _snap_ to a given geometry.

### Rationale for the enhancement or information for reproducing the error
This would allow a circular domain approximated by an n-gon to be adapted to an m-gon (with m>n) rather than a more refined n-gon.

### Identified impact
Added capability.

Ping @friedmud, who talked about this years ago already. I had a use case for this and implemented it using GeneralUserObjects and the MeshChangedInterface...

### PR DESCRIPTION
**Pull request by dschwen**
Closes #9587

### CODE REVIEW

File: framework/include/userobject/GeometrySphere.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+#include "GeometryBase.h"
+
+class GeometrySphere;
+
+template <>
+InputParameters validParams<GeometrySphere>();
+
+/**
+ * Snaps teh selected nodes to the surface of a sphere (or circular disk in 2D)
```
Comment (permcody) [inline framework/include/userobject/GeometrySphere.h]: teh

**Review by permcody — COMMENTED** [review]
Does this code work with distributed mesh? I see a call to `node_ref` in the main algorithm, and it only starts with one element so I'm not sure. I'd be nice if it did.

File: framework/include/userobject/GeometrySphere.h (hunk 1 above)
Comment (dschwen) [inline framework/include/userobject/GeometrySphere.h]: Ggggnnnnnnaaaaaarrrrrggh!!!!!!

**Comment by dschwen** [comment]
> Does this code work with distributed mesh? I see a call to `node_ref` in the main algorithm

Good question. I think the answer hinges on 

```_mesh.getNodeList(boundary_id)``` 

returning the _right  thing_ <sup>TM</sup> for distributed meshes.

**Comment by permcody** [comment]
more_elems

Did you try this with nx/ny>1? It seems that you end up with a block of undistorted elements and then a ring of distorted elements around that. I suppose this is fine but it seems like we should be able to bias every node in this simple problem?

**Comment by permcody** [comment]
Alternatively, we could simply just not allow people to set nx/ny when using this user object. I wanted to see if distributed mesh worked. It appears that at least works.

**Comment by dschwen** [comment]
@permcody this is exactly how this is expected to work (only the boundary nodes get moved)! The test is an exaggerated application. Ideally your coarse mesh is a bit closer to the desired geometry than a square is to a circle!

**Comment by friedmud** [comment]
I agree with @dschwen .

We can always also add support for the mesh smoothers present in libMesh ( http://libmesh.github.io/doxygen/classlibMesh_1_1MeshSmoother.html )... but for transient problems you also need to do remapping (including for material properties) so it gets tricky.

**Comment by permcody** [comment]
Countdown to user, actually using this new BASE CLASS in a production run in 100... 99...

image

**Review by permcody — APPROVED** [review]

**Pull request merged**